src/daemons/rhsmcertd.c

```diff
@@ -244,34 +247,154 @@ long long gen_random(long long max) {
+ * @return Return true, when uploading was successful. Otherwise, return false.
+ */
+static gboolean
+upload_package_profile ()
+{
+    gboolean ret;
+    const char * argv[] = {PACKAGE_PROFILE_UPLOADER, NULL};
+    gchar *standard_output = NULL;
+    gchar *standard_error = NULL;
+    gint wait_status;
+    GError *error = NULL;
+
+    debug ("Spawning new process of uploading package profile...");
+
+    // Following function will block until process is finished. Other signals
+    // are queued and will be processed, when following process is finished.
+    // It means other signals (SIGTERM) are blocked too. systemd can resolve
+    // this issue. When sending SIGTERM does not terminate daemon, then it
+    // sends SIGKILL signal after some timeout.
+    ret = g_spawn_sync(
+            "/",                // Working directory
+            (gchar**)argv,      // Executable with arguments
+            NULL,               // List of environments is read from parent process
+            G_SPAWN_DEFAULT,    // Default flags
+            NULL,               // No function running before spawning process
+            NULL,               // No argument for such function is needed as well
+            &standard_output,   // stdout if any
+            &standard_error,    // stderr if any
+            &wait_status,       // status if any
+            &error);            // Error if any
+
+    debug ("Uploading of package profile finished");
+
+    if (standard_output != NULL && strlen(standard_output) > 0) {
+        debug ("stdout of uploading package profile: %s", standard_output);
+    }
+    if (standard_error != NULL && strlen(standard_error) > 0) {
+        debug ("stderr of uploading package profile: %s", standard_error);
+    }
+
+    // Error is usually not NULL, when it wasn't possible to span child process
+    // for some reason (e.g. file does not exist)
+    if (error != NULL) {
+        error ("Spanning child process (uploading profile) failed: %s", error->message);
+        return false;
+    }
+
+    ret = WIFEXITED (wait_status);
+    if (wait_status != 0) {
+        ret = WIFSIGNALED(wait_status);
+        if (ret == true) {
+            int signal_num = WTERMSIG(wait_status);
+            error ("Uploading of package profile was terminated by a signal: %d", signal_num);
+            return false;
+        }
+        if (ret != true) {
+            error ("Uploading of package profile was terminated with status: %d", wait_status);
+            return false;
         }
-        info ("rhsmcertd is shutting down...");
-        signal (signo, SIG_DFL);
-        raise (signo);
     }
+    ret = WEXITSTATUS(wait_status);
+    if (ret == true) {
+        info ("Uploading of package profile performed successfully.");
+        return true;
+    } else {
+        warn ("Uploading of package profile terminated with status code: %d", wait_status);
+        return false;
+    }
```

Comment:
now that I read more the glib API (and its source), i think there is a nice function that can greatly help us here: `g_spawn_check_exit_status()` (or `g_spawn_check_exit_status()`, but we cannot use it as it is too new); according to its doc, it should be a simple matter of doing:
```c
  if (g_spawn_check_exit_status (wait_status, &error) == FALSE) {
    warn ("Uploading of package profile failed: %s", error ? error->message : "<unknown>");
  }
  info ("Uploading of package profile performed successfully.");
```
no more ugly C stdlib API :)

Reply:
The `g_spawn_check_exit_status()` is considered as deprecated since 2.70 and `g_spawn_check_wait_status()` should be used. On the other hand `g_spawn_check_wait_status()` is not available at least on RHEL8. I think that current implementation just works on all RHELs. I can see that `g_spawn_check_wait_status()` has only one benefit. It is platform independent, but it is something that we don't need.

Reply:
`g_spawn_check_wait_status()` is simply a better name for `g_spawn_check_exit_status()`; the old function does not do anything more than calling the new one:
https://gitlab.gnome.org/GNOME/glib/-/blob/4f406f3de63e19e7abc7416e69ea5348bea48a50/glib/gspawn.c#L1227

More than portability, it does the right checks for the exit status; in particular:
```c
    ret = WIFEXITED (wait_status);
    if (wait_status != 0) {
```
this check is not correct, as `wait_status` contains bits regarding the exit/wait status (eg it was signalled, etc); relying on that `wait_status` equal to 0 means "exited without issues" is undefined behaviour
```c
        ret = WIFSIGNALED(wait_status);
        if (ret == true) {
```
`WIFSIGNALED()` does bitshift internally; even without knowing that, all you can assume is that it returns a boolean-like value, rather than a direct boolean
```c
    ret = WEXITSTATUS(wait_status);
    if (ret == true) {
```
this "works" because of implicit boolean conversion, as anything different than 0 is true; better handle it as value, rather than boolean
```c
        if (ret == true) {
            ...
        }
        if (ret != true) {
```
the second `if` is basically redundant, and it is better replaced by an `else`